[GPU] [CoreIPC] Improve expression validation in WebCore::SVGFilterRenderer

zakariaridouh opened the pull request:
#### d49bc45cdcaa2538b30fce4c01a5980f381ba674
<pre>
[GPU] [CoreIPC] Improve expression validation in WebCore::SVGFilterRenderer
<a href="https://bugs.webkit.org/show_bug.cgi?id=300468">https://bugs.webkit.org/show_bug.cgi?id=300468</a>
<a href="https://rdar.apple.com/161641790">rdar://161641790</a>

Reviewed by Said Abou-Hallawa.

We can sometimes get into a state where we have an empty `expressions` vector with a
non empty `effects` vector when making an SVGFilterRenderer, which should not be the case.

This causes us to crash when trying to get the last valid expression in `WebCore::SVGFilterRenderer::apply`.

Fix by hardening the IPC validation function.

* LayoutTests/ipc/empty-svgfilterrenderer-expression-crash-expected.txt: Added.
* LayoutTests/ipc/empty-svgfilterrenderer-expression-crash.html: Added.
* Source/WebCore/svg/graphics/filters/SVGFilterRenderer.cpp:
(WebCore::SVGFilterRenderer::apply):
(WebCore::SVGFilterRenderer::isValidSVGFilterExpression):

Canonical link: <a href="https://commits.webkit.org/301540@main">https://commits.webkit.org/301540@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0b5c2589d0593a76af458a0ccefb75e49584d2d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45877 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36694 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132999 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77995 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/92d15166-b208-41a4-969b-d7b8d9dedd97) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128092 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46541 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54423 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96101 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64209 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5111a7ff-6ab6-4da1-8cc5-1a7641421240) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129169 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37233 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112898 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76583 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/962e3d31-c539-492a-9514-ed07d651650e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36133 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31077 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76471 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107015 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135700 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52970 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40686 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104598 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53432 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109122 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104301 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26609 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49738 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28072 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50353 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52870 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58701 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52186 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55532 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53903 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on LayoutTests/ipc/empty-svgfilterrenderer-expression-crash.html:
```diff
@@ -0,0 +1,149 @@
+            renderingBackendIdentifier,
+        );
+        remoteRenderingBackend.CreateImageBuffer({
+            logicalSize: { width: 1.1210387714598537e-43, height: 0 },
```
Creating this ImageBuffer should fail. Please file a bug to make `ImageBufferCGPDFDocumentBackend::create()` validate its parameters.

zakariaridouh replied:
Tracked here: [GPU] ImageBufferCGPDFDocumentBackend::create() should validate its parameters
https://bugs.webkit.org/show_bug.cgi?id=300487

Will open a patch for it after this one lands

shallawa commented on LayoutTests/ipc/empty-svgfilterrenderer-expression-crash.html:
```diff
@@ -0,0 +1,149 @@
+            logicalSize: { width: 1.1210387714598537e-43, height: 0 },
+            renderingMode: 2,
+            renderingPurpose: 0,
+            resolutionScale: -6.119062772890424e21,
```
I do not think the source `ImageBuffer` should be created with these weird parameters to reproduce the bug. I think the bug does not need PDF backend with negative resolutionFactor to reproduce. Can these parameters be replaced with normal parameters like:
```
    logicalSize: { width: 100, height: 1000 },
    renderingMode: 1,
    renderingPurpose: 0,
    resolutionScale: 1,
```

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilterRenderer.cpp:
```diff
@@ -314,6 +314,9 @@ RefPtr<FilterImage> SVGFilterRenderer::apply(FilterImage* sourceImage, FilterRes
     ASSERT(filterRenderingModes().contains(FilterRenderingMode::Software));
 
+    if (m_expression.isEmpty())
+        return nullptr;
```
I think this condition can't be true. `SVGFilterRenderer` can't be created if the expression is empty.  This is guaranteed by `SVGFilterRenderer::create()` and `ArgumentCoder<WebCore::SVGFilterRenderer>::decode()`. Maybe we can add something like
```
ASSERT(isValidSVGFilterExpression(m_expression, m_effects));
```

zakariaridouh committed:
[GPU] [CoreIPC] Improve expression validation in WebCore::SVGFilterRenderer
https://bugs.webkit.org/show_bug.cgi?id=300468
rdar://161641790

Reviewed by Said Abou-Hallawa.

We can sometimes get into a state where we have an empty `expressions` vector with a
non empty `effects` vector when making an SVGFilterRenderer, which should not be the case.

This causes us to crash when trying to get the last valid expression in `WebCore::SVGFilterRenderer::apply`.

Fix by hardening the IPC validation function.

* LayoutTests/ipc/empty-svgfilterrenderer-expression-crash-expected.txt: Added.
* LayoutTests/ipc/empty-svgfilterrenderer-expression-crash.html: Added.
* Source/WebCore/svg/graphics/filters/SVGFilterRenderer.cpp:
(WebCore::SVGFilterRenderer::apply):
(WebCore::SVGFilterRenderer::isValidSVGFilterExpression):

Canonical link: https://commits.webkit.org/301540@main
